File: backend/app/domain/entities/payment.py
"""Entidades de dominio para pagos operativos de servicios (BE-011)."""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class PaymentMethod(str, Enum):
    """Metodos de pago operativo."""

    CASH = "cash"
    TRANSFER = "transfer"
    CARD = "card"
    OTHER = "other"


class PaymentStatus(str, Enum):
    """Estados de un pago operativo."""

    PAID = "paid"
    CANCELLED = "cancelled"


class Payment(BaseModel):
    """Pago operativo de servicio (salida completa)."""

    model_config = ConfigDict(from_attributes=True)

    id: int | None = None
    appointment_id: int = Field(..., gt=0)
    service_id: int = Field(..., gt=0)
    clinic_id: int = Field(..., gt=0)
    amount: int = Field(..., ge=0)
    method: PaymentMethod
    amount_received: int | None = Field(None, ge=0)
    change_amount: int | None = Field(None, ge=0)
    status: PaymentStatus
    paid_at: datetime
    cancelled_at: datetime | None = None
    created_by: int | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class PaymentCreate(BaseModel):
    """Datos de entrada para registrar un pago (caso de uso)."""

    appointment_id: int = Field(..., gt=0)
    service_id: int = Field(..., gt=0)
    amount: int = Field(..., ge=0)
    method: PaymentMethod
    amount_received: int | None = Field(None, ge=0)
    created_by: int | None = None

File: frontend/src/shared/api/payment.ts
import { apiClient } from './client';

export type PaymentMethod = 'cash' | 'transfer' | 'card' | 'other';

export type PaymentStatus = 'paid' | 'cancelled';

export interface Payment {
  id: number;
  appointment_id: number;
  service_id: number;
  clinic_id: number;
  amount: number;
  method: PaymentMethod;
  amount_received: number | null;
  change_amount: number | null;
  status: PaymentStatus;
  paid_at: string;
  cancelled_at: string | null;
  created_by: number | null;
}

export interface PaymentCreateData {
  appointment_id: number;
  service_id: number;
  amount: number;
  method: PaymentMethod;
  /** Obligatorio y >= amount cuando method === 'cash'. */
  amount_received?: number | null;
}

export interface PaymentListParams {
  page?: number;
  page_size?: number;
  appointment_id?: number;
  from_date?: string;
  to_date?: string;
  status?: PaymentStatus;
}

export interface PaymentPageMeta {
  page: number;
  page_size: number;
  total: number;
  pages: number;
}

export interface PaymentListResponse {
  items: Payment[];
  meta: PaymentPageMeta;
}

const BASE_PATH = '/payments';

// El backend espera datetime ISO para from_date/to_date. Los inputs de la UI
// (type="date") emiten solo fecha "YYYY-MM-DD"; se normaliza añadiendo hora.
function toIsoDate(value: string, endOfDay: boolean): string {
  if (/^\d{4}-\d{2}-\d{2}$/.test(value)) {
    return `${value}T${endOfDay ? '23:59:59' : '00:00:00'}`;
  }
  return value;
}

export async function createPayment(data: PaymentCreateData): Promise<Payment> {
  return apiClient.post<Payment>(BASE_PATH, data);
}

export async function getPayment(paymentId: number): Promise<Payment> {
  return apiClient.get<Payment>(`${BASE_PATH}/${paymentId}`);
}

export async function cancelPayment(paymentId: number): Promise<Payment> {
  return apiClient.post<Payment>(`${BASE_PATH}/${paymentId}/cancel`);
}

export async function listPayments(
  params: PaymentListParams = {},
): Promise<PaymentListResponse> {
  const queryParams = new URLSearchParams();
  queryParams.set('page', String(params.page ?? 1));
  queryParams.set('page_size', String(params.page_size ?? 20));

  if (params.appointment_id) {
    queryParams.set('appointment_id', String(params.appointment_id));
  }
  if (params.from_date) {
    queryParams.set('from_date', toIsoDate(params.from_date, false));
  }
  if (params.to_date) {
    queryParams.set('to_date', toIsoDate(params.to_date, true));
  }
  if (params.status) {
    queryParams.set('status', params.status);
  }

  const query = queryParams.toString();
  return apiClient.get<PaymentListResponse>(
    query ? `${BASE_PATH}?${query}` : BASE_PATH,
  );
}
